### ISSUE TITLE
Support for optional version numbering

### ISSUE THREAD

**Issue by kvesteri** [sqlalchemy/sqlalchemy#5283]
I have a project where certain models are required to have [version numbering](https://docs.sqlalchemy.org/en/13/orm/versioning.html) for certain API endpoints. These API endpoints are used by the application's offline functionalities for conflict management.

The vast majority of the application works only online and I'd like to leave the myriad API endpoints untouched. Preferably I could turn the version numbering on / off easily. Something along the lines of

```python
session.versioning = False

with session.enable_versioning():
    some_model.some_column = 3 
    session.commit()
```

I've been trying to figure out a workaround and dig in deep in the SQLAlchemy internals but couldn't figure out a solutions for this. Have people had similar requirements? Or any ideas for a workaround?

**Comment by zzzeek** [comment]
OK first pass at understanding the problem:

1. you say that "certain models" need to have versioning, but then your code seems to indicate that it is "certain **requests**" that need to have versioning, based on what you're doing with the session

2. "certain models" also implies that you are talking about "certain mappers", meaning, those mappers have the version_id_col thing, and the others dont, so I think your phrase "certain models" is not actually what you mean

3. so let's assume it's "certain requests", and based what I see, maybe you are saying, you want to use the ORM and flush objects, and then in only certain requests, bump the version number ?     OK, that is just, set the value of a column for that flush.

4. it's not clear if you are looking just for a way to have a number incrementing sometimes, which is not a big deal, or you want the actual concurrency check that versioning is intended for to occur but only on certain flushes?     It may be the latter since as you've mentioned you dug into the internals, and you likely saw the logic for the "needs_version_id" boolean which determines whether or not this version id is tested and it is true, there is not an easy way for some sessions to check the row count and other sessions to not do so, for the same model.

if you just want a number that increments sometimes and you don't care about the concurrency check feature, then I would just accomplish version counting using another kind of counting mechanism.    if you actually want concurrency checks to happen for only some sessions for the same model within the same process (e.g. you can't just change the mapper for some processes), that might be a difficult feature right now as that is not an assumed use case for the included versioning feature.

**Comment by zzzeek** [comment]
TLDR the concurrency check feature is the only special thing you get from version_id_col and there's probably not a way to turn that on/off per session.    for just an incrementing number there's a lot of ways to do that, use mapper before_update() events, etc.

**Comment by zzzeek** [comment]
you can also maybe do this at the model level.    that is, map two different classes to the same table, and one has the versioning.   this wont play nicely with relationships but might be all you need.    you can *maybe* have these two mappers / classes in an inheritance relationship also , like implicitly single table inheritance, which might allow them to work more smoothly in more complex situations like relationships.

**Comment by zzzeek** [comment]
OK here's outer-space edge land, the mapper keeps the identity key of two classes separate if there is no "polymorphic identity" set up, i might have to confirm that is tested, but that means we can map two classes to the same table in an inheritance relationship, which means they can probably be referred towards by the same relationship(), here's a demo:

```
from sqlalchemy import *
from sqlalchemy.orm import *
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.ext.declarative import declared_attr

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)
    version_id_col = Column(Integer, default=1)


class AVersioned(A):
    __mapper_args__ = {"version_id_col": A.__table__.c.version_id_col}


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

s.add(A())
s.commit()


objv = s.query(AVersioned).first()
objv.data = "new data"
s.commit()

assert objv.version_id_col == 2

obj = s.query(A).first()
obj.data = "even newer data"
s.commit()

assert obj.version_id_col == 2

assert obj is not objv
```

**Comment by zzzeek** [comment]
it's tested, wow https://github.com/sqlalchemy/sqlalchemy/blob/master/test/orm/inheritance/test_basic.py#L1163

**Comment by kvesteri** [comment]
Thanks for the example @zzzeek ! With the inheritance I can simply use the inherited mapper in certain special endpoints and then leave all the other myriad endpoints as they are.

**Comment by zzzeek** [comment]
ok that works?    if so we can close.  I'm making some really massive changes to SQLAlchemy right now which you'll probably find interesting but it's taking all my attention away from everything else.

**Comment by kvesteri** [comment]
> ok that works? if so we can close. I'm making some really massive changes to SQLAlchemy right now which you'll probably find interesting but it's taking all my attention away from everything else.

Yep it works :+1: Thanks again
